### PR TITLE
XMeans

### PR DESCRIPTION
**Pull request by Komzpa**

### CODE REVIEW

File: liblwgeom/lwkmeans.c [hunk 1]
```diff
@@ -202,11 +361,74 @@ kmeans_init(POINT4D *objs, uint32_t n, POINT4D *centers, uint32_t k)
...
+	uint8_t converged = LW_FALSE;
+	assert(max_k >= min_k);
+	assert(max_k >= 2);
+	uint32_t cur_k = 2;
+
+	kmeans_init(objs, n, centers, cur_k);
+
+	for (uint32_t t = 0; t < KMEANS_MAX_ITERATIONS; t++)
+	{
+		for (uint32_t i = 0; i < KMEANS_MAX_ITERATIONS; i++)
```
Comment (HanwGeek) [inline liblwgeom/lwkmeans.c:393]: ### 
```suggestion
              // Naive kmeans for `improve params`
		for (uint32_t i = 0; i < KMEANS_MAX_ITERATIONS; i++)
```

File: liblwgeom/lwkmeans.c [hunk 2]
```diff
@@ -29,8 +40,171 @@ distance3d_sqr_pt4d_pt4d(const POINT4D *p1, const POINT4D *p2)
...
+		  uint8_t we_had_a_split)
+{
+	POINT4D *temp_objs = lwalloc(sizeof(POINT4D) * n);
+	int *temp_clusters = lwalloc(sizeof(int) * n);
+	double *temp_radii = lwalloc(sizeof(double) * n);
+	POINT4D *temp_centers = lwalloc(sizeof(POINT4D) * n);
+
+	uint32_t new_k = k;
+
+	for (int cluster = 0; cluster < (int)k; cluster++)
```
Comment (HanwGeek) [inline liblwgeom/lwkmeans.c:109]:
```suggestion
      // Try to split every cluster into two and check its score
	for (int cluster = 0; cluster < (int)k; cluster++)
```

Comment (HanwGeek) [inline liblwgeom/lwkmeans.c:109]:
```suggestion
        // Try to split each cluster into two and check their score
	for (int cluster = 0; cluster < (int)k; cluster++)
```

**Comment by Komzpa** [comment]
Add documentation and unit tests

**Pull request closed without merging**